File: Source/Runned.cs
/* 
@author: Gael Lopes Da Silva
@project: Runned
@github: https://github.com/Gael-Lopes-Da-Silva/Runned
@gitlab: https://gitlab.com/Gael-Lopes-Da-Silva/Runned
*/

using System.Diagnostics;
using System.Runtime.InteropServices;

class Runned
{
    private static bool exitCode = false;
    private static bool noOutput = false;
    private static bool timeDetail = false;
    private static bool showInput = false;

    private static void WriteColoredLine(ConsoleColor color, string info, string message)
    {
        Console.ForegroundColor = color;
        Console.Write($"{info.ToUpper()}: ");
        Console.ResetColor();
        Console.WriteLine(message);
    }

    private static void ExecuteCommand(string commands)
    {
        Stopwatch timer = new();
        string systemConsole = string.Empty;
        string processCommands = string.Empty;

        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            systemConsole = "CMD.exe";
            processCommands = $"/c {commands}";
        }
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
        {
            systemConsole = "/bin/bash";
            processCommands = $"-c \"{commands}\"";
        }

        timer.Start();
        using Process process = new();
        process.StartInfo = new ProcessStartInfo(systemConsole, processCommands)
        {
            CreateNoWindow = true,
            UseShellExecute = false,
            RedirectStandardInput = true,
            RedirectStandardOutput = true
        };
        process.Start();
        process.WaitForExit();
        timer.Stop();

        string input = commands.Trim('\n');
        string output = process.StandardOutput.ReadToEnd();

        if (output == string.Empty) output = "None";

        if (!noOutput)
        {
            Console.WriteLine("---- Output ----");
            Console.WriteLine($"{output.Trim('\n')}");
            Console.WriteLine("----------------\n");
        }

        if (!timeDetail)
        {
            WriteColoredLine(ConsoleColor.Green, "time", $"{timer.Elapsed:ss\\.fffff}");
        }
        else
        {
            WriteColoredLine(ConsoleColor.Green, "time", $"{timer.Elapsed}");
        }

        if (exitCode) WriteColoredLine(ConsoleColor.DarkGray, "exitcode", $"{process.ExitCode}");
        if (showInput) WriteColoredLine(ConsoleColor.DarkGray, "input", $"{input}");
    }

    private static string ExtractArguments(string[] args)
    {
        string arguments = string.Empty;

        foreach (string arg in args)
        {
            switch (arg.ToLower())
            {
                case "--exitcode":
                case "-ec":
                    exitCode = true;
                    break;

                case "--nooutput":
                case "-no":
                    noOutput = true;
                    break;

                case "--timedetail":
                case "-td":
                    timeDetail = true;
                    break;

                case "--showinput":
                case "-si":
                    showInput = true;
                    break;

                default:
                    arguments += $"{arg} ";
                    break;
            }

        }

        return arguments.TrimEnd(' ');
    }

    public static void Main(string[] args)
    {
        if (args.Length > 0)
        {
            try
            {
                ExecuteCommand(ExtractArguments(args));
            }
            catch (Exception error)
            {
                WriteColoredLine(ConsoleColor.Red, "error", $"The following problem has occured: {error.Message}");
            }
        }
        else
        {
            Console.WriteLine("Runned v0.1");
            Console.WriteLine("A simple tool to check execution time of commands.\n");

            WriteColoredLine(ConsoleColor.DarkGray, "usage", "Runned.exe [<your commands>]\n");

            WriteColoredLine(ConsoleColor.DarkGray, "options", "");
            Console.WriteLine("  -ec --exitcode    Display the exit code of the executed command.");
            Console.WriteLine("  -no --nooutput    Disable the output.");
            Console.WriteLine("  -td --timedetail  Add more detail to the elapsed time.");
            Console.WriteLine("  -si --showinput   Display the input given by the user.");
        }
    }
}

File: source/runned.nim
#[
@author: Gael Lopes Da Silva
@project: Runned
@github: https://github.com/Gael-Lopes-Da-Silva/Runned
@gitlab: https://gitlab.com/Gael-Lopes-Da-Silva/Runned
]#

import std/[os, terminal, strutils, times]

var
  exitCode: bool = false
  showInput: bool = false

proc echoColorLine(messageType: string, message: string, color: ForegroundColor, bright: bool): void =
  stdout.setForegroundColor(color, bright)
  stdout.write(messageType.toUpper())
  stdout.resetAttributes()
  stdout.write(": " & message & '\n')

proc echoColor(messageType: string, color: ForegroundColor, bright: bool): void =
  stdout.setForegroundColor(color, bright)
  stdout.write(messageType.toUpper())
  stdout.resetAttributes()
  stdout.write(":\n")

proc executeCommand(commands: string): int =
  let
    timerStart: float = cpuTime()
    returnValue: int = execShellCmd(commands)
    timerEnd: float = cpuTime() - timerStart
  
  echoColorLine "\ntime", $timerEnd & " s", fgGreen, false

  if exitCode: echoColorLine "exitcode", $returnValue, fgBlack, true
  if showInput: echoColorLine "input", commands.strip(trailing = true, chars = Newlines), fgBlack, true

  return returnValue

proc extractArguments(args: seq[string]): string =
  var arguments: string = ""

  for arg in args:
    case arg.toLower():
    of "--exitcode", "-ec", "-e": exitCode = true
    of "--showinput", "-si", "-i": showInput = true
    else: arguments.add(arg & " ")

  return arguments.strip(trailing = true)

proc main(args: seq[string]): int =
  if (len(args) > 0):
    try:
      return executeCommand(extractArguments(args))
    except Exception:
      echoColorLine "error", "The following problem has occured: " & getCurrentExceptionMsg(), fgRed, false
  else:
    echo "Runned v0.2"
    echo "A simple tool to check execution time of commands.\n"
    echoColorLine "usage", "Runned.exe <your commands>\n", fgBlack, true
    echoColor "options", fgBlack, true
    echo "  -e -ec --exitcode    Display the exit code of the executed command."
    echo "  -i -si --showinput   Display the input given by the user."

  return QuitSuccess

when isMainModule: quit main(commandLineParams())
